### ISSUE TITLE
Shut down during VTune profiling

### ISSUE THREAD

**Issue by jiseongg** [intel/mkl-dnn#549]
Hi, I'm studying the CPU performance with CNN computation. Especially, I'm interested in AVX acceleration while using PyTorch. I found that PyTorch AVX acceleration is supported by `mkl-dnn`, which is located at `pytorch/third_party/ideep/`. And I succeeded to debug the source codes inside this directory.

However, when I tried profiling with VTune amplifier, it literally shut down. (Not only the VTune, but the whole computer system stops.) And I'm supposing that it may shut down at the `mkl-dnn` codes. And it runs well with single thread, but only shuts down with multi threads. So, I built `mkl-dnn` again with the command below. (At first time, I built the `mkl-dnn` with default option)

```bash
mkdir build && cd build
cmake -DCMAKE_BUILD_TYPE=DEBUG -DCMAKE_C_COMPILER=icc -DCMAKE_CXX_COMPILER=icpc -DCMAKE_Fortran_COMPILER=ifort -DDNNL_ENABLE_JIT_PROFILING=ON ..
```
It is to make `libmkldnn.so` linked to `libiomp5.so`, not `libgomp5.so` for OpenMP profiling with VTune. And I verified this result after I built it. Now I'm going to try profiling with VTune again, but I'm worried because the server I'm using is shared with others. So I want to ask a question here whether the problem I mentioned above was because of the `libmkldnn.so` and `libgomp5.so` library, because I'm not sure.

Also, there is another reason I'm hesitating. I found some interesting log when I enter `cmake` command.

```bash
-- VTune profiling environment is unset
```

Is there anyone who tried profiling the program including `mkl-dnn` with VTune Amplifier?

-----------------------------------------------------------------------------

### Environment
DNNL includes hardware-specific optimizations and may behave
differently on depending on the compiler and build environment. Include
the following information to help reproduce the issue:
* CPU make and model
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                16
On-line CPU(s) list:   0-15
Thread(s) per core:    1
Core(s) per socket:    8
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 62
Model name:            Intel(R) Xeon(R) CPU E5-2650 v2 @ 2.60GHz
Stepping:              4
CPU MHz:               1200.252
CPU max MHz:           3400.0000
CPU min MHz:           1200.0000
BogoMIPS:              5201.92
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              20480K
NUMA node0 CPU(s):     0,2,4,6,8,10,12,14
NUMA node1 CPU(s):     1,3,5,7,9,11,13,15
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm cpuid_fault pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase smep erms xsaveopt dtherm ida arat pln pts md_clear flush_l1d

* OS version (`uname -a`)
4.15.0-60-generic #67~16.04.1-Ubuntu SMP

* Compiler version (`gcc --version`)
icc (ICC) 19.0.0.117 20180804

* MKLROOT value (`echo MKLROOT=$MKLROOT`)
no result

* CMake version (`cmake --version`)
cmake version 3.5.1

* CMake output log
build worked well. One message I'm interested in is below.

```bash
-- VTune profiling environment is unset
```

**Comment by vpirogov** [comment]
Hi @jiseongg,

The behavior you describe might be a result of a conflict between two OpenMP runtimes. Having `libiomp5` and `libgomp` loaded at the same time leads to undefined behavior and might result in incorrect results or a segfault.

When you build the library with Intel compiler it has runtime dependency on libiomp5. You can verify that by running `ldd libkmldnn.so`. If Pytorch loads `libgomp` and `libmkldnn` you'll get undefined behavior. There are two ways around that:
* Build the library with GNU compiler and `libgomp`.
* Make sure that Pytorch is using `libiomp5` as well. One option to do that might work is  `LD_PRELOAD=libiomp5.so`, but I'm not completely sure about that.

The error message regarding VTune you see indicates that build system was not able to find VTune installation and link to libraries necessary to talk to VTune. You need to set `VTUNEROOT` build option. See [here](http://intel.github.io/mkl-dnn/v0/perf_profile.html). Note, that for v1.0 and above (current master) VTune works out of the box and does not require special steps at build time.

**Comment by rsdubtso** [comment]
I think the system crash is actually due to VTune drivers and has little to do with MKL-DNN (esp. since the VTune integration seems to be disabled). I would recommend upgrading VTune if possible...

**Comment by jiseongg** [comment]
@vpirogov, 
If I do the first way you mentioned, that environment may be what I've already tested. At first, `libmkldnn.so` and `libtorch.so` both loaded `libgomp.so` and VTune made the system shut down. So I made a `libmkldnn.so` load `libiomp.so` by building with Intel compiler. But after reading your comment, I feel a little strange. If two libraries both loads `libgomp.so`, why it had a problem?

I think I have to try the second way you mentioned, making sure that PyTorch is using `libiomp5.so`. I will report it after it is done.

And about the error message regarding VTune. VTune version I'm using is `vtune_amplifier_2019.0.2.570779` (the directory name). Are you saying the message about VTune I mentioned may not be a problem?


@rsdubtso,
Thank you for your suggestion. I will search about it and report here.

**Comment by vpirogov** [comment]
Considering that the system shut down completely I believe Roma is right and it has more to do with VTune that Intel MKL-DNN.

The message you see indicates that Intel MKL-DNN build system was not able to locate VTune. You need to set VTUNEROOT to point to the location if VTune installation.

**Comment by vpirogov** [comment]
@jiseongg, do you need any other help?

**Comment by jiseongg** [comment]
@vpirogov, I've finally solved the problem, and now I'm using VTune well.

I failed to build PyTorch with Intel Compiler. I've built it with `gcc` and my PyTorch code is loading `libgomp.so`, after `libtorch.so` and `libmkl_xxx.so` are loaded, respectively. But I've learned that if I'm not focusing on OpenMP region, it's not quite necessary to link with `libiomp5.so`. What is important in my study is mkl-dnn's AVX implementation. I'll note the environment setting and build command below.

1. I've set environment variable of VTune root directory
In [mkl-dnn/profiling.cmake](https://github.com/intel/mkl-dnn/blob/0125f28c61c1f822fd48570b4c1066f96fcb9b2e/cmake/profiling.cmake)
I added
```cmake
set(VTUNEROOT "/path/to/vtune_amplifier_version")
```
(I think it can be done by just `export` command and setting it globally.)

2. Build PyTorch with following command.

```bash
DEBUG=1 USE_CUDA=0 USE_MKLDNN=1 MKLDNN_THREADING=OMP python setup.py develop
```

It is enough to profile about AVX in mkl-dnn.
